src/ripple/rpc/handlers/GatewayBalances.cpp

```diff
@@ -187,43 +194,41 @@ Json::Value doGatewayBalances (RPC::Context& context)
 
     if (! sums.empty())
     {
-        Json::Value& j = (result [jss::obligations] = Json::objectValue);
+        Json::Value j;
         for (auto const& e : sums)
         {
             j[to_string (e.first)] = e.second.getText ();
         }
+        result [jss::obligations] = std::move (j);
     }
 
-    if (! hotBalances.empty())
-    {
-        Json::Value& j = (result [jss::balances] = Json::objectValue);
-        for (auto const& account : hotBalances)
+    auto populate = [](
+        std::map <AccountID, std::vector <STAmount>> const& array,
+        Json::Value& result,
+        Json::StaticString const& name)
         {
-            Json::Value& balanceArray = (j[to_string (account.first)] = Json::arrayValue);
-            for (auto const& balance : account.second)
+            if (!array.empty())
             {
-                Json::Value& entry = balanceArray.append (Json::objectValue);
-                entry[jss::currency] = to_string (balance.issue ().currency);
-                entry[jss::value] = balance.getText();
+                Json::Value j;
+                for (auto const& account : array)
+                {
+                    Json::Value balanceArray;
```

Comment:
Our JSON library has some interesting semantics for how null values can be treated as empty objects or empty arrays. You pretty much can't ignore these semantics because they affect a lot of the library's behavior, so I'm not sure it's necessarily a bad thing to use them here.